File: problem/revcomp/solution/php/1.php
<?php
#
# The Computer Language Benchmarks Game
# https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
#
# reverse complement in PHP
# contributed by Danny Sauer
#

# Weʼll need some definitions
define( ʼLINE_LENGTHʼ, 60 );
define( ʼSRCʼ, ʼCGATMKRYVBHDʼ);
define( ʼDSTʼ, ʼGCTAKMYRBVDHʼ);
$str = ʼʼ;
$seq = ʼʼ;

# read in the file, a line at a time
while( !feof(STDIN) ) {
    $str = trim(fgets(STDIN));
    if( $str && $str[0] == ʼ>ʼ ){
        # if weʼre on a comment line, print the previous seq and move on
        print_seq();
        echo $str, "\n";
    }else{
        # otherwise, just append to the sequence
        $seq .= $str;
    }
}
print_seq();

exit;

# print the sequence out, if it exists
function print_seq(){
    global $seq; # no time-consuming argument passing for us! :)
    if($seq != ʼʼ){
        echo wordwrap( strrev( strtr(strtoupper($seq), SRC, DST) ),
                       LINE_LENGTH, "\n", true ), "\n";
    }
    $seq = ʼʼ;
}
?>

File: problem/revcomp/solution/python3/1.py
# The Computer Language Benchmarks Game
# https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
#
# contributed by Matt Vollrath

from sys import stdout


COMPLEMENTS = bytes.maketrans(
    bʼACGTUMRWSYKVHDBNacgtumrwsykvhdbnʼ,
    bʼTGCAAKYWSRMBDHVNTGCAAKYWSRMBDHVNʼ,
)
COMMENT = ord(ʼ>ʼ)


def reverse_sequence(sequence):
    chunk = bytearray()
    complemented = sequence.translate(COMPLEMENTS, bʼ\nʼ)
    seq_len = len(complemented)
    last_line_len = seq_len % 60
    if last_line_len:
        chunk += bʼ\nʼ + complemented[:last_line_len]
    for i in range(last_line_len, seq_len, 60):
        chunk += bʼ\nʼ + complemented[i:i+60]
    return chunk[::-1]


def generate_sequences(lines):
    heading = None
    sequence = bytearray()
    for line in lines:
        if line[0] == COMMENT:
            if heading:
                yield heading, sequence
                sequence = bytearray()
            heading = line
        else:
            sequence += line
    yield heading, sequence


if __name__ == ʼ__main__ʼ:
    stdin = open(0, buffering=1)

    for heading, sequence in generate_sequences(stdin.buffer):
        stdout.buffer.write(heading)
        stdout.buffer.write(reverse_sequence(sequence))
        stdout.buffer.flush()

